src/text_render.rs

```diff
@@ -60,7 +60,18 @@ impl TextRenderer {
         let resolution = viewport.resolution();
 
         for text_area in text_areas {
-            for run in text_area.buffer.layout_runs() {
+            let is_run_visible = |run: &cosmic_text::LayoutRun| {
+                let start_y = (text_area.top + run.line_top) as i32;
+                let end_y = (text_area.top + run.line_top + run.line_height) as i32;
+
+                start_y <= text_area.bounds.bottom && text_area.bounds.top <= end_y
```

Comment:
Does this handle lines that are partially visible? For example, lines that only have their top half or bottom half inside the text area.

If I understand correctly, this seems to be checking if the line is fully bounded by the text area, but we'd want to handle any lines overlapping/intersecting the text bounds instead.

Reply:
> Does this handle lines that are partially visible? For example, lines that only have their top half or bottom half inside the text area.
> 
> If I understand correctly, this seems to be checking if the line is fully bounded by the text area, but we'd want to handle any lines overlapping/intersecting the text bounds instead.

yes this should check whether the line's y range intersects with the visual bounds' y range. it confuses me too tbh

i based it on the `intersects` code from emath: https://github.com/emilk/egui/blob/64d5f948e43b4be93e3a4541c87dcce74c948df4/crates/emath/src/range.rs#L140-L142

Reply:
Ah ok I didn't notice the top and bottom bounds were switched. Sounds good!